### ISSUE TITLE
Consider consolidating conda solves in CI tests

### ISSUE THREAD

**Issue by charlesbluca** [rapidsai/build-planning#22]
Currently (using cuML as an example here), the conda test environment initialization for most CI jobs looks something like creating the test environment:

```bash
rapids-dependency-file-generator \
  --output conda \
  --file_key test_python \
  --matrix "cuda=${RAPIDS_CUDA_VERSION%.*};arch=$(arch);py=${RAPIDS_PY_VERSION}" | tee env.yaml

rapids-mamba-retry env create --force -f env.yaml -n test
```

And then downloading and installing build artifacts from previous jobs on top of this environment:

```
CPP_CHANNEL=$(rapids-download-conda-from-s3 cpp)
PYTHON_CHANNEL=$(rapids-download-conda-from-s3 python)

...

rapids-mamba-retry install \
  --channel "${CPP_CHANNEL}" \
  --channel "${PYTHON_CHANNEL}" \
  libcuml cuml
```

In addition to forcing us to eat the cost of a second conda environment solve, in many cases this can cause some pretty drastic changes to the environment which can be blocking - for example, consider this [cuML run](https://github.com/rapidsai/cuml/actions/runs/7992518167/job/21827898950?pr=5761) which fails because conda is unable to solve a downgrade from Arrow 15.0.0 ([build 5](https://github.com/conda-forge/arrow-cpp-feedstock/pull/1314/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9a)) to 14.0.1.

Our current workaround for this is to manually add pinnings to the testing dependencies initially solved such that the artifact installation can be solved, but this can introduce a lot of burden in needing to:

- identify what packages/changes are blocking artifact installation
- open PR(s) modifying the impacted repos
- follow up on each impacted repo to potentially remove the pinning later on

Would it be possible to consolidate some (or all) of these conda environment solves by instead:

1. downloading the conda artifacts before creating the environment
1. updating the dependencies.yaml (or, if this isn't possible, the generated conda environment file) to include the desired packages, making sure to explicitly specify source channel to ensure we're picking up the build artifacts
1. creating the environment with this patched file

In my mind, the main blocker I could see to this working would be if `rapids-download-conda-from-s3` requires some conda packages contained in the testing environment to work.

EDIT: Updating to capture state of other projects:

- [x] rmm(https://github.com/rapidsai/rmm/pull/1824, https://github.com/rapidsai/rmm/pull/1828)
- [x] raft(https://github.com/rapidsai/raft/pull/2580, https://github.com/rapidsai/raft/pull/2587)
- [x] cudf(https://github.com/rapidsai/cudf/pull/17995, https://github.com/rapidsai/cudf/pull/18014)
- [x] kvikio(https://github.com/rapidsai/kvikio/pull/513, https://github.com/rapidsai/kvikio/pull/636)
- [x] cuml(https://github.com/rapidsai/cuml/pull/5781, https://github.com/rapidsai/cuml/pull/6321)
- [x] cucim (https://github.com/rapidsai/cucim/pull/833, https://github.com/rapidsai/cucim/pull/835)
- [x] cugraph(https://github.com/rapidsai/cugraph/pull/4935)
- [x] cugraph-gnn(https://github.com/rapidsai/cugraph-gnn/pull/144)
- [x] cumlprims_mg(no changes needed!)
- [x] cuspatial(https://github.com/rapidsai/cuspatial/pull/1387)
- [x] cuvs(https://github.com/rapidsai/cuvs/pull/684, https://github.com/rapidsai/cuvs/pull/701, https://github.com/rapidsai/cuvs/pull/745)
- [x] cuxfilter(https://github.com/rapidsai/cuxfilter/pull/663, https://github.com/rapidsai/cuxfilter/pull/664)
- [x] dask-cuda(https://github.com/rapidsai/dask-cuda/pull/1448, https://github.com/rapidsai/dask-cuda/pull/1452)
- [x] nx-cugraph(https://github.com/rapidsai/nx-cugraph/pull/90)
- [x] rapids-cmake(no changes needed!)
- [x] rapids-dask-dependency(no changes needed!)
- [x] ucx-py(https://github.com/rapidsai/ucx-py/pull/1101)
- [x] ucxx(https://github.com/rapidsai/ucxx/pull/373, https://github.com/rapidsai/ucxx/pull/375)
- [x] wholegraph(no changes needed! superseded by cugraph-gnn)

**Comment by msarahan** [comment]
I don't fully understand how much you need to be rebuilding, but if the answer is "not much" then maybe you could use https://conda.github.io/conda-pack/ or consider building a docker image instead of re-creating the environment each time?

**Comment by bdice** [comment]
I am proposing a flag `--prepend-channels` for `rapids-dependency-file-generator` that will enable a one-step conda environment creation. This would let us use local channels with CI artifacts and would eliminate the second conda solve used to install CI artifacts. https://github.com/rapidsai/dependency-file-generator/pull/67

**Comment by bdice** [comment]
> consider building a docker image instead of re-creating the environment each time?

This is a possible solution, but we'd need to maintain a separate Docker image for every repository and every test matrix configuration (e.g. cuDF Python tests on ARM using Python 3.9 and CUDA 12.2). I don't think we have the infrastructure to manage that while _also_ ensuring that the conda environments are regularly updated (every night?) for all the possible containers so that we are aware of upstream changes that affect us. The workflows we have today would require an explosively large number of large containers (many of the containers would need to contain an entire CTK installed by conda).

**Comment by raydouglass** [comment]
The docker repo suffers from a similar problem and it was (partly) mitigated by dynamically editing the environment file: https://github.com/rapidsai/docker/blob/branch-24.04/context/notebooks.sh#L31, so just throwing that out there.

**Comment by vyasr** [comment]
Once https://github.com/rapidsai/cuml/pull/5781 is merged, do we plan to roll out the same changes to the rest of RAPIDS? IIUC that's what it would take to close this issue.

**Comment by jakirkham** [comment]
Seems reasonable

We may want ensure that we are always getting cached CI packages too, which would close out issue: https://github.com/rapidsai/build-planning/issues/14

**Comment by bdice** [comment]
@vyasr Yes, my plan is to convert all of RAPIDS to this single-solve approach.

@jakirkham The approach I'm taking in https://github.com/rapidsai/cuml/pull/5781 will pin the `major.minor` versions of the packages as shown below. That isn't a full solution for #14 because nightlies of the same `major.minor` could still be selected, but it should help constrain the solver and reduce the likelihood of problems. If you have further suggestions please weigh in on #14 or on that cuML PR (if the code suggestions are small) and I'll adopt your proposals in a follow-up change.

```
  test_libcuml:
    common:
      - output_types: conda
        packages:
          - libcuml==24.4.*
          - libcuml-tests==24.4.*
  test_cuml:
    common:
      - output_types: conda
        packages:
          - libcuml==24.4.*
          - cuml==24.4.*
```

**Comment by jakirkham** [comment]
Ok it sounds like that is mostly an issue when we are getting close to a release. Is that right?

Agree that could be an issue, but that is a much smaller issue than picking up the prior release (especially well after it went out)

**Comment by bdice** [comment]
Yeah, I think the problematic case would be something like changing the dependencies of a package in a way that the conda packages from CI artifacts are not compatible with the environment, and the solver chooses recent nightlies instead of the PR artifacts. @charlesbluca proposed we could pin the exact channels like `${CPP_CHANNEL}::libcuml` but we'd have to find a way to insert that channel information into the `dependencies.yaml` since the channel isn't known ahead of time.

**Comment by vyasr** [comment]
An additional minor benefit is that this will speed up CI by reducing the number of solves.

**Comment by bdice** [comment]
> An additional minor benefit is that this will speed up CI by reducing the number of solves.

This might not have a large impact on CI runtime. See https://github.com/rapidsai/cudf/pull/15201#issuecomment-1973562102. tl;dr, test startup time is dominated by network activity (downloading multiple GB of packages) and the potential conda solve time savings are comparatively small.

**Comment by msarahan** [comment]
Taking this on with high priority because cuspatial CI is blocked, and this might help there.

**Comment by bdice** [comment]
@msarahan Is there a plan to continue on this work? We have implemented it for cuml and cuspatial but nothing else. It would be good to have alignment across RAPIDS to prevent future issues.

**Comment by msarahan** [comment]
Added task list to parent issue. We'll work on this as time allows.

**Comment by KyleFromNVIDIA** [comment]
Closed as all listed repos have had this implemented.

**Comment by jakirkham** [comment]
Woohoo! 🎊 🎉 

Thanks Kyle for all the PRs and everyone for help reviewing 🙏

**Comment by jakirkham** [comment]
Kyle did we only need the cuVS PR ( https://github.com/rapidsai/cuvs/pull/745 )? Or are there other PRs still needed?

**Comment by KyleFromNVIDIA** [comment]
I reopened it only for that cuvs PR. Closing now.
